Add the global nesting capability

guoqing-noaa opened the pull request:
# Description

This PR adds the global nesting capability to the global workflow so that one can run a regional domain nested in a global run.

Resolves #2544 

# Type of change
<!-- Delete all except one -->
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- Clone and build on Hera
- Forecast-only on Hera

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

guoqing-noaa committed:
Add the global nesting capability

aerorahul commented on ush/parsing_model_configure_FV3.sh:
```diff
@@ -59,6 +59,27 @@ if [[ ! -f ${template} ]]; then
 rm -f "${DATA}/model_configure"
 atparse < "${template}" >> "${DATA}/model_configure"
+
+if [[ "${DO_NEST:-NO}" == YES ]] ; then
+    cat >> model_configure <<EOF
+
+# Output settings for the nest.
+<output_grid_02>
+output_grid:             ${NEST_OUTPUT_GRID}
+imo:                     ${npx_nest}
+jmo:                     ${npy_nest}
+cen_lon:                 ${target_lon}
+cen_lat:                 ${target_lat}
+lon1:                    ${NEST_LON1}
+lat1:                    ${NEST_LAT1}
+lon2:                    ${NEST_LON2}
+lat2:                    ${NEST_LAT2}
+dlon:                    ${NEST_DLON}
+dlat:                    ${NEST_DLAT}
+</output_grid_02>
+EOF
+fi
```
We should not do this.  The ufs-weather-model should have a `model_configure` for `nest` case, that should be used.

guoqing-noaa replied:
Thanks! I will make necessary changes here and create a PR to the ufs-weather-model as well.

aerorahul commented on parm/config/gfs/config.resources:
```diff
@@ -600,6 +600,13 @@ case ${step} in
         [[ "${DO_ICE}" == "YES" ]] && nthreads_cice6=1
       fi
 
+      if [[ ${ntiles} -gt 6 ]] ; then
```
```suggestion
      if (( ntiles > 6 )); then
```

aerorahul commented on parm/config/gfs/config.ufs:
```diff
@@ -68,6 +68,21 @@ if [[ "${skip_mom6}" == "false" ]] || [[ "${skip_cice6}" == "false" ]] || [[ "${
   skip_mediator=false
 fi
 
+if [[ "${DO_NEST:-NO}" == YES ]] ; then
```
```suggestion
if [[ "${DO_NEST:-NO}" == "YES" ]] ; then
    # Describe nest location, interaction with parent, etc.
```

aerorahul commented on parm/config/gfs/config.ufs:
```diff
@@ -105,6 +120,24 @@ case "${fv3_res}" in
         export WRTTASK_PER_GROUP_PER_THREAD_PER_TILE=1
         export WRITE_GROUP_GFS=1
         export WRTTASK_PER_GROUP_PER_THREAD_PER_TILE_GFS=1
+        if [[ ${DO_NEST:-NO} == YES ]] ; then
+            export DELTIM=450
+            export layout_x_gfs=4
+            export layout_y_gfs=4
+            export layout_x_nest=12
+            export layout_y_nest=10
+            export nest_refine=4
+            export nest_ioffset=4
+            export nest_joffset=9
+            export npx_nest=361
+            export npy_nest=241
+            export NEST_DLON=0.25
+            export NEST_DLAT=0.25
+            export WRITE_GROUP=2
+            export WRTTASK_PER_GROUP_PER_THREAD_PER_TILE=2
+            export WRITE_GROUP_GFS=2
+            export WRTTASK_PER_GROUP_PER_THREAD_PER_TILE_GFS=2
+        fi
```
If you could please cast this as:
```bash
if [[ "${DO_NEST:-NO}" == "YES" ]]; then
   settings with nest
else
    settings without nest
fi
```
that would help ensuring they are grouped appropriately

guoqing-noaa replied:
Sure. Thanks!

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -175,6 +175,10 @@ EOF
       if [[ "${WRITE_DOPOST}" == ".true." ]]; then
         ${NLN} "${COM_ATMOS_MASTER}/${RUN}.t${cyc}z.master.grb2f${FH3}" "GFSPRS.GrbF${FH2}"
         ${NLN} "${COM_ATMOS_MASTER}/${RUN}.t${cyc}z.sfluxgrbf${FH3}.grib2" "GFSFLX.GrbF${FH2}"
+        if [[ "${DO_NEST:-NO}" == YES ]] ; then
+            ${NLN} "${COM_ATMOS_MASTER}/${RUN}.t${cyc}z.nest.grb2f${FH3}" "GFSPRS.GrbF${FH2}.nest02"
```
The indent is 2 spaces in this script.  Please try to stay consistent.
4 spaces is fine, but when the strings are long, every little bit helps with readability.

guoqing-noaa replied:
Sure. Thanks!

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -196,7 +200,12 @@ FV3_nml() {
+      FV3_namelists nest
+  else
+      FV3_namelists #default to global without nest
+  fi
```
```suggestion
  # Call the appropriate namelist functions
  if [[ "${DO_NEST:-NO}" == YES ]] ; then
      FV3_namelists global_with_nest
      FV3_namelists nest
  else
      FV3_namelists
  fi
```

aerorahul commented on parm/config/gfs/config.base:
```diff
@@ -184,6 +184,13 @@ export DO_AERO="NO"
+  export ntiles=7
+  export NEST_OUTPUT_GRID="regional_latlon"
+else
+  export ntiles=6
```
It probably is a good idea to put `export ntiles=6` in `parm/config/gefs/config.base` as well as `export DO_NEST="NO"`

guoqing-noaa replied:
Sure. Thanks!

aerorahul commented on parm/config/gfs/config.fcst:
```diff
@@ -253,7 +253,11 @@ export DO_SPPT=${DO_SPPT:-"NO"}
+  export DO_CA="NO" #CA does not work with nesting. 
+else
+  export DO_CA=${DO_CA:-"YES"}
+fi
```
```suggestion
export DO_CA=${DO_CA:-"YES"}
if [[ "${DO_NEST:-NO}" == "YES" ]] ; then
  export DO_CA="NO"  # CA does not work with nesting.
fi
```

aerorahul commented on parm/config/gfs/config.stage_ic:
```diff
@@ -38,4 +38,8 @@ case "${CASE}" in
 esac
 
+if [[ "${DO_NEST:-NO}" == "YES" ]] ; then
+  export CPL_ATMIC=GLOBAL-NEST_${CASE}
```
```suggestion
  export CPL_ATMIC="GLOBAL-NEST_${CASE}"
```

aerorahul commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -6,6 +6,20 @@
 # shellcheck disable=SC2034
 FV3_namelists(){
 
+# First argument tells us which namelist we're writing:
```
We are in the process of replacing this with using a template from the ufs-weather-model, and this change makes it very difficult for us to do that at this time.
I will discuss with @junwang-noaa to find out if there is a way to do it before this part of the PR.

guoqing-noaa replied:
Sounds good. Thanks!

aerorahul reviewed (changes requested):
Most of the changes look good.
There are some minor changes requested.
2 major comments:
1. `model_configure` template should be used from ufs-weather-model rather than `cat` 
2. `FV3_namelists` is about to be replaced with a much smaller, leaner method that uses `input.nml.IN` from ufs-weather-model.  The changes in this PR make it difficult to address.  We should discuss offline to come up with a mitigation strategy so that it does not prevent either of us to do the things we want.

guoqing-noaa committed:
Apply suggestions from code review

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guoqing-noaa committed:
Apply suggestions from code review

guoqing-noaa committed:
minor fix

guoqing-noaa committed:
minor changes

guoqing-noaa commented:
> Most of the changes look good. There are some minor changes requested. 2 major comments:
> 
> 1. `model_configure` template should be used from ufs-weather-model rather than `cat`
> 2. `FV3_namelists` is about to be replaced with a much smaller, leaner method that uses `input.nml.IN` from ufs-weather-model.  The changes in this PR make it difficult to address.  We should discuss offline to come up with a mitigation strategy so that it does not prevent either of us to do the things we want.

@aerorahul Thanks a lot for your great help and great suggestions. I think I addressed most comments except `FV3_namelists`. 
Are you good that I go ahead to set up a tag up meeting with you and relevant parties on this?

guoqing-noaa commented:
FYI, PR https://github.com/ufs-community/ufs-weather-model/pull/2252 was created to add tests/parm/model_configure_nest.IN

guoqing-noaa committed:
remove self.config.ntiles=6 in analysis.py

guoqing-noaa committed:
revert changes in parsing_namelists_FV3.sh

guoqing-noaa committed:
add a new file - parsing_namelists_FV3_nest.sh

guoqing-noaa committed:
update parsing_namelists_FV3_nest.sh for global nest

guoqing-noaa committed:
update config.ufs

guoqing-noaa committed:
update parsing_namelists_FV3_nest.sh

guoqing-noaa committed:
update forecast_postdet.sh

guoqing-noaa committed:
add quotation mark to 'YES'

guoqing-noaa committed:
disable 'masking return value' warnings

guoqing-noaa committed:
shellcheck disable=SC2312

guoqing-noaa committed:
Merge branch 'develop' into do_nest

aerorahul reviewed (approved):
Looks good.
Thanks for accepting the suggestions and working them in.
Just a note:  
A model update will be needed to find the `model_configure_nest.IN` at some point.  Until then, the `DO_NEST=YES` feature will need to use a ufs-weather-model hash provided by someone else.

guoqing-noaa committed:
Merge branch 'develop' into do_nest

guoqing-noaa commented:
Thanks, @aerorahul !

Yes, at the moment, we will manually put in place the `model_configure_nest.IN` file when we run the global nest workflow.

guoqing-noaa committed:
Merge branch 'develop' into do_nest

aerorahul committed:
Merge branch 'develop' into do_nest

WalterKolczynski-NOAA reviewed (approved):
Conditionally approved pending successful completion of CI tests.

guoqing-noaa commented:
Thanks, @WalterKolczynski-NOAA and @aerorahul !
